### PR TITLE
notes

### PR DESCRIPTION
**Pull request by ManickaP**
cc: @geoffkizer  @CarnaViire @wfurt

### CODE REVIEW

**Comment by wfurt** [comment]
thanks for taking notes @ManickaP. Should? we merge it to keep it with the code?

**Comment by ManickaP** [comment]
> Should? we merge it to keep it with the code?

I'll transform the comment like notes into real comments and the ones which are pointing out problems into issues. We should not merge this PR. I'll make a different one. Let's not pollute our code base with meeting notes.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 1]
```diff
@@ -50,6 +56,7 @@ private sealed class State
             // set when ReadState.PendingRead:
             public Memory<byte> ReceiveUserBuffer;
             public CancellationTokenRegistration ReceiveCancellationRegistration;
+            // CR: Potentially we could get rid of back reference to the parent object since we don't care about keeping them alive while async msquic call is happening.
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:56]: #55103

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 2]
```diff
@@ -65,12 +72,16 @@ private sealed class State
             // Resettable completions to be used for multiple calls to send.
             public readonly ResettableCompletionSource<uint> SendResettableCompletionSource = new ResettableCompletionSource<uint>();
 
+            // CR: We should get rid off it, we're not interested in it.
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:72]: #55704

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 3]
```diff
@@ -65,12 +72,16 @@ private sealed class State
...
             public ShutdownWriteState ShutdownWriteState;
 
             // Set once writes have been shutdown.
+            // CR: We should get rid of this
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:76]: #55704

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 4]
```diff
@@ -65,12 +72,16 @@ private sealed class State
...
             public readonly TaskCompletionSource ShutdownWriteCompletionSource = new TaskCompletionSource(TaskCreationOptions.RunContinuationsAsynchronously);
 
+            // CR: Shutdown should be merged into CloseAsync.
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:79]: #756

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 5]
```diff
@@ -113,16 +124,21 @@ internal MsQuicStream(MsQuicConnection.State connectionState, SafeMsQuicStreamHa
             }
             catch
             {
+                // CR: we should release the handle here as well as we do in outbound. Does int interfere with this being run on msquic thread? Ask NICK.
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:124]: #55814

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 6]
```diff
@@ -113,16 +124,21 @@ internal MsQuicStream(MsQuicConnection.State connectionState, SafeMsQuicStreamHa
...
+                // Should we return a specific error code in MsQuicConnection to clean the handle for us?
                 _state.StateGCHandle.Free();
                 throw;
             }
 
+            // CR: We should move this before callback handler registration.
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:130]: #55815

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 7]
```diff
@@ -113,16 +124,21 @@ internal MsQuicStream(MsQuicConnection.State connectionState, SafeMsQuicStreamHa
...
             if (!connectionState.TryAddStream(this))
             {
                 _state.StateGCHandle.Free();
                 throw new ObjectDisposedException(nameof(QuicConnection));
             }
 
+            // CR: We should set it before SetCallbackHandlerDelegate since we can get CONNETCION_CLOSED event and that needs this.
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:137]: #55815

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 8]
```diff
@@ -170,6 +186,7 @@ internal MsQuicStream(MsQuicConnection.State connectionState, QUIC_STREAM_OPEN_F
                 throw;
             }
 
+            // CR: The same as with the inbound.
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:186]: #55815

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 9]
```diff
@@ -191,6 +208,7 @@ internal MsQuicStream(MsQuicConnection.State connectionState, QUIC_STREAM_OPEN_F
 
         internal override bool CanRead => _disposed == 0 && _canRead;
 
+        // CR: Might be useful to introduce Is(Uni|Bi)Directional to distinguish the type of the stream instead of using Can*.
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:208]: #55816

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 10]
```diff
@@ -220,6 +238,7 @@ internal override ValueTask WriteAsync(ReadOnlySequence<byte> buffers, Cancellat
 
         internal override async ValueTask WriteAsync(ReadOnlySequence<byte> buffers, bool endStream, CancellationToken cancellationToken = default)
         {
+            // CR: we should propagate exceptions up here for write state.
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:238]: @wfurt I don't exactly remember what the problem was, could you remind me before I create an issue?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 11]
```diff
@@ -258,6 +277,8 @@ internal override async ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, bool e
 
         private CancellationTokenRegistration HandleWriteStartState(CancellationToken cancellationToken)
         {
+            // CR: we should make the flow as we do in Read, where we check the state and make transition in a lock as a first thing
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:277]: #55817

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 12]
```diff
@@ -286,6 +307,8 @@ private CancellationTokenRegistration HandleWriteStartState(CancellationToken ca
             }
 
             // if token was already cancelled, this would execute synchronously
+            // CR: we need to track the registration and dispose it even in case of exception thrown outside of this method.
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:307]: #55437

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 13]
```diff
@@ -318,6 +341,7 @@ private CancellationTokenRegistration HandleWriteStartState(CancellationToken ca
                         throw new QuicStreamAbortedException(_state.SendErrorCode);
                     }
 
+                    // CR: this should be some other exception, we should throw OCE unless CT has been fired.
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:341]: #55619

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 14]
```diff
@@ -555,6 +579,7 @@ private void StartShutdown(QUIC_STREAM_SHUTDOWN_FLAGS flags, long errorCode)
             QuicExceptionHelpers.ThrowIfFailed(status, "StreamShutdown failed.");
         }
 
+        // CR: We should get rid of this
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:579]: #55704

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 15]
```diff
@@ -1146,6 +1171,8 @@ private static void CleanupSendState(State state)
         }
 
         // TODO prevent overlapping sends or consider supporting it.
+        // CR: Why do we have 3 version of sending? We should unite as much as possible.
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:1175]: #55818

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 16]
```diff
@@ -1200,6 +1227,7 @@ private static void CleanupSendState(State state)
             return _state.SendResettableCompletionSource.GetTypelessValueTask();
         }
 
+        // CR: for kestrel
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:1231]: #55818

**Comment by ManickaP** [comment]
And comments are here: https://github.com/dotnet/runtime/pull/55819

**Pull request closed without merging**